pkg/controllers/jobset_controller.go

```diff
@@ -15,45 +15,299 @@ package controllers
+			return err
+		}
+
+		// Skip the job if it is already active or succeeded.
+		skip, err := r.shouldSkipJob(ctx, job)
+		if err != nil {
+			return err
+		}
+		if skip {
+			klog.Infof("skipping job %s", job.Name)
+			continue
+		}
+
+		klog.Infof("creating job %s", job.Name)
+
+		// Job must be created before headless service so that the job.spec.selector
+		// field is populated.
+		if err := r.Create(ctx, job); err != nil {
+			klog.Error(err, "unable to create Job for JobSet", "job", job)
+			return err
+		}
+
+		// Next, create headless service if specified for this job.
+		if jobTemplate.Network.EnableDNSHostnames != nil && *jobTemplate.Network.EnableDNSHostnames {
+			if err := r.createHeadlessSvcIfNotExist(ctx, req, jobSet, job); err != nil {
+				klog.Infof("error creating headless service: %v", err)
+				return err
+			}
+		}
+
+		klog.Info("created Job for JobSet", "job", job)
+	}
+	return nil
+}
+
+// TODO: look into adopting service and updating the selector
+// if it is not matching the job selector.
+func (r *JobSetReconciler) createHeadlessSvcIfNotExist(ctx context.Context, req ctrl.Request, jobSet *jobsetv1alpha.JobSet, job *batchv1.Job) error {
+	// Check if service already exists. Service name is same as job name.
+	var headlessSvc corev1.Service
+	namespacedSvc := types.NamespacedName{Namespace: jobSet.Namespace, Name: job.Name}
+
+	if err := r.Get(ctx, namespacedSvc, &headlessSvc); err != nil {
+
+		// Get updated job object from apiserver which will have the selector populated.
+		namespacedJob := types.NamespacedName{Name: job.Name, Namespace: jobSet.Namespace}
+
+		// Need to retry in case the freshly created job object has not been persisted quite yet.
+		if err := retry.OnError(
+			defaultRetry,
+			func(error) bool { return true }, // always retry
+			func() error {
+				if err := r.Get(ctx, namespacedJob, job); err != nil {
+					return err
+				}
+				return nil
+			},
+		); err != nil {
+			return err
+		}
+
+		// Construct headless service defintion.
+		selectorMap, err := metav1.LabelSelectorAsMap(job.Spec.Selector)
+		if err != nil {
+			return err
+		}
+		headlessSvc := corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      job.Name,
+				Namespace: req.Namespace,
+			},
+			Spec: corev1.ServiceSpec{
+				ClusterIP: "None",
+				Selector:  selectorMap,
+			},
+		}
+
+		// set controller owner reference for garbage collection and reconcilation
+		if err := ctrl.SetControllerReference(jobSet, &headlessSvc, r.Scheme); err != nil {
+			klog.Error(err, "error setting controller owner reference for headless service", "service", headlessSvc)
+			return err
+		}
+
+		// Create headless service.
+		if err := r.Create(ctx, &headlessSvc); err != nil {
+			klog.Error(err, "unable to create headless service", "service", headlessSvc)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *JobSetReconciler) shouldSkipJob(ctx context.Context, job *batchv1.Job) (bool, error) {
```

Comment:
this function is used in the context of job create; for now, since we are not yet implementing the recreate the semantics, the only reason we should be creating a job is if it didn't already exist.

Reply:
Alright, I simplified the logic to only check if the job exists or not and act accordingly. I will wait until we implement recreate semantics to add handling based on job status, etc.